src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -514,63 +514,74 @@ object messages {
   }
 
   case class WrongNumberOfArgs(fntpe: Type, argKind: String, expectedArgs: List[TypeParamInfo], actual: List[untpd.Tree])(implicit ctx: Context)
-    extends Message(18) {
+    extends Message(19) {
     val kind = "Syntax"
     val expectedCount = expectedArgs.length
     val actualCount = actual.length
     val msgPrefix = if (actualCount > expectedCount) "Too many" else "Not enough"
 
-    val expectedArgString = (fntpe.widen match {
-      case pt: MethodOrPoly => pt //ensure we return a type that will have useful typeParms
-      case _ => fntpe
-    }).typeParams.map(_.paramName.show.split("\\$").last).mkString("[", ", ", "]")
+    val expectedArgString = fntpe.widen.typeParams.map(_.paramName.unexpandedName.show).mkString("[", ", ", "]")
+
+    val actualArgString = actual.map(_.show).mkString("[", ", ", "]")
 
-    val actualArgString = actual.map(_.show.split("\\.").last).mkString("[", ", ", "]")
+    val prettyName = fntpe.termSymbol match {
+      case NoSymbol => fntpe.show
+      case symbol   => symbol.showFullName
+    }
```

Comment:
I would like to make another PR to add def prettyName (or something like it) to a few types.  I will leave it as a follow up unless you prefer it as part of this PR.

With these changes
```
object Test {
  var bad = List[Int, String]
  case class Bar[D, E](baz: D, qux: E)
  var bad3: Bar[Int] = null
}
```
results in
```
-- [E019] Syntax Error: test.scala -------------------------------------------------------------------------------------
16 |  var bad = List[Int, String]
   |            ^^^^^^^^^^^^^^^^^
   |            Too many type arguments for scala.collection.immutable.List.apply[A]
   |            expected: [A]
   |            actual:   [Int, String]

Explanation
===========
You have supplied too many type parameters

For example List takes a single type parameter (List[A])
If you need to hold more types in a list then you need to combine them
into another data type that can contain the number of types you need,
In this example one solution would be to use a Tuple:

val tuple2: (Int, String) = (1, "one")
val list: List[(Int, String)] = List(tuple2)

-- [E019] Syntax Error: test.scala -------------------------------------------------------------------------------------
19 |  var bad3: Bar[Int] = null
   |            ^^^^^^^^
   |            Not enough type arguments for Test.Bar[D, E]
   |            expected: [D, E]
   |            actual:   [Int]

Explanation
===========
You have not supplied enough type parameters
If you specify one type parameter then you need to specify every type parameter.
```